libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,588 @@
+ *
+ * @param[in] packetType PUBACK, PUBREC, PUBREL, or PUBCOMP.
+ * @param[in] opType Send, or Receive.
+ *
+ * @return `true` if corresponds to outgoing publish, else `false`.
+ */
+static bool _isPublishOutgoing( MQTTPubAckType_t packetType,
+                                MQTTStateOperation_t opType );
+
+/**
+ * @brief Find a packet ID in the state record.
+ *
+ * @param[in] records State record array.
+ * @param[in] recordCount Length of record array.
+ * @param[in] packetId packet ID to search for.
+ * @param[out] pQos QoS retrieved from record.
+ * @param[out] pCurrentState state retrieved from record.
+ *
+ * @return index of the packet id in the record if it exists, else the record length.
+ */
+static size_t _findInRecord( const MQTTPubAckInfo_t * records,
+                             size_t recordCount,
+                             uint16_t packetId,
+                             MQTTQoS_t * pQos,
+                             MQTTPublishState_t * pCurrentState );
+
+/**
+ * @brief Store a new entry in the state record.
+ *
+ * @param[in] records State record array.
+ * @param[in] recordCount Length of record array.
+ * @param[in] packetId, packet ID of new entry.
+ * @param[in] qos QoS of new entry.
+ * @param[in] publishState state of new entry.
+ *
+ * @return MQTTSuccess, MQTTNoMemory, MQTTStateCollision.
+ */
+static MQTTStatus_t _addRecord( MQTTPubAckInfo_t * records,
+                                size_t recordCount,
+                                uint16_t packetId,
+                                MQTTQoS_t qos,
+                                MQTTPublishState_t publishState );
+
+/**
+ * @brief Update and possibly delete an entry in the state record.
+ *
+ * @param[in] records State record array.
+ * @param[in] recordCount Length of record array.
+ * @param[in] recordIndex index of record to update.
+ * @param[in] newState New state to update.
+ * @param[in] shouldDelete Whether an existing entry should be deleted.
+ */
+static void _updateRecord( MQTTPubAckInfo_t * records,
+                           size_t recordCount,
+                           size_t recordIndex,
+                           MQTTPublishState_t newState,
+                           bool shouldDelete );
+
+
+static bool _validateTransitionPublish( MQTTPublishState_t currentState,
+                                        MQTTPublishState_t newState,
+                                        MQTTStateOperation_t opType,
+                                        MQTTQoS_t qos )
+{
+    bool isValid = false;
+    switch( currentState )
+    {
+        case MQTTStateNull:
+            /* Transitions from invalid occur when storing a new entry into the record. */
+            if( opType == MQTT_RECEIVE )
+            {
+                isValid = ( newState == MQTTPubAckSend ) || ( newState == MQTTPubRecSend );
+            }
+            break;
+        case MQTTPublishSend:
+            /* Outgoing publish. */
```

Comment:
Perhaps add to the comment that all outgoing publishes start in this state due to the reserve operation.

Reply:
Updated comment.